### PR TITLE
Add tanh_cuda support for complex types

### PR DESCRIPTION
**Pull request by krshrimali**
Similar to #38400, this PR atempts to add support for complex types for `tanh` (CUDA).

cc: @anjali411

### CODE REVIEW

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu [hunk 1]
```diff
@@ -150,15 +152,22 @@ void cosh_kernel_cuda(TensorIterator& iter) {
...
   return ::tanh(v);
 }
 
+template<typename T>
+__host__ __device__ static inline thrust::complex<T> tanh_wrapper(thrust::complex<T> v) {
+  return thrust::tanh(v);
+}
+
 void tanh_kernel_cuda(TensorIterator& iter) {
   AT_DISPATCH_FLOATING_TYPES_AND2(ScalarType::Half, ScalarType::BFloat16, iter.dtype(), "tanh_cuda", [&]() {
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:167]: `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2`

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -246,6 +246,7 @@ def method_tests():
         ('log2', uniform_scalar(1e-2j, requires_grad=True), NO_ARGS, 'complex_scalar', (True,)),
         ('tanh', (S, S, S), NO_ARGS, '', (True,)),
         ('tanh', (), NO_ARGS, 'scalar', (True,)),
+        ('tanh', (S, S, S), NO_ARGS, 'complex', (True,)),
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:249]: you should include `tanh` in this [list](https://github.com/pytorch/pytorch/blob/7f11079769b0c9cb678a1046f1b0c9259ca3a3ac/test/test_autograd.py#L4193)

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu (hunk 1 above)
Comment (krshrimali) [reply aten/src/ATen/native/cuda/UnaryGeometricKernels.cu]: Sorry, my bad. Missed this. Should be resolved here: https://github.com/pytorch/pytorch/pull/38486/commits/a930fce4ae0637e5a178a0c124efa9ffd98670f7. Thanks!

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:249]: Do you mean [here](https://github.com/pytorch/pytorch/blob/c06bd9289c9b73b200a4275a6ea6c34e8129ed4a/test/test_autograd.py#L4193)? I think I already included `tanh` in the list. Please correct me if I'm getting anything wrong here. Thanks!

Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:249]: yeah sorry my b for some reason that change didn't show up for me last time

**Review by anjali411 — APPROVED** [review]

**Comment by anjali411** [comment]
I see test failure for test_autograd `RuntimeError: "tanh_backward_cpu" not implemented for 'ComplexDouble'` so you would need to update `tanh_backward_kernel` as well for the autograd test to run fine for complex

**Review by anjali411 — CHANGES_REQUESTED** [review]
let's update the `tanh_backward_kernel` and `tanh_backward_kernel_cuda`

**Comment by krshrimali** [comment]
Thanks @anjali411! Didn't realize that it was "backward" (missed the keyword there). I've pushed the changes, can you please review the PR again?

**Comment by anjali411** [comment]
LGTM there's a merge conflict so just rebase it and should be fine

**Comment by krshrimali** [comment]
@anjali411 - can you please review the changes? Please let me know if there is anything left for this PR. :)

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu [hunk 3]
```diff
@@ -150,15 +152,22 @@ void cosh_kernel_cuda(TensorIterator& iter) {
...
+template<typename T>
+__host__ __device__ static inline thrust::complex<T> tanh_wrapper(thrust::complex<T> v) {
+  return thrust::tanh(v);
+}
+
 void tanh_kernel_cuda(TensorIterator& iter) {
-  AT_DISPATCH_FLOATING_TYPES_AND2(ScalarType::Half, ScalarType::BFloat16, iter.dtype(), "tanh_cuda", [&]() {
-    AT_SKIP_BFLOAT16_IF_NOT_ROCM(scalar_t, "tanh_cuda", [&] {
-      gpu_kernel(iter, []GPU_LAMBDA(scalar_t a) -> scalar_t {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(ScalarType::Half, ScalarType::BFloat16, iter.dtype(), "tanh_cuda", [&]() {
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:167]: `AT_DISPATCH_FLOATING_AND_C10_COMPLEX_TYPES_AND2`

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu [hunk 4]
```diff
@@ -150,15 +152,22 @@ void cosh_kernel_cuda(TensorIterator& iter) {
...
+    using thrust_t = typename ztype_cuda<scalar_t>::thrust_t;
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:168]: we can remove this once the above change is made.

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu [hunk 5]
```diff
@@ -150,15 +152,22 @@ void cosh_kernel_cuda(TensorIterator& iter) {
...
 }
 
+// We manually override tanh because std::tanh does not work with thrust::complex types.
 template<typename scalar_t>
 __host__ __device__ static inline scalar_t tanh_wrapper(scalar_t v) {
   return ::tanh(v);
 }
 
+template<typename T>
+__host__ __device__ static inline thrust::complex<T> tanh_wrapper(thrust::complex<T> v) {
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:162]: `__host__ __device__ static inline c10::complex<T> tanh_wrapper(c10::complex<T> v)`

**Comment by anjali411** [comment]
left some minor comments.
test failures look flaky but let's rebase and see the CI

**Comment by krshrimali** [comment]
Screwed things up while rebasing. Attempting to fix this.

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu (hunk 4 above)
Comment (krshrimali) [reply aten/src/ATen/native/cuda/UnaryGeometricKernels.cu]: Done!

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu (hunk 5 above)
Comment (krshrimali) [reply aten/src/ATen/native/cuda/UnaryGeometricKernels.cu]: Done!

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu (hunk 3 above)
Comment (krshrimali) [reply aten/src/ATen/native/cuda/UnaryGeometricKernels.cu]: I don't see this macro defined in the master branch. Have added this macro though.

**Comment by krshrimali** [comment]
Things should be back to normal now. 

* Changes to `tanh_backward_kernel` (cpu) have been ommitted since a PR adding these changes (#37791) was merged recently.
* Changes to `tanh_backward_kernel_cuda` are done, but using macro: `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES`. Please let me know if this needs to be changed.
* We are now using `c10::complex` and appropriate macros for `c10::complex` in `aten/src/ATen/native/cuda/UnaryGeometricKernels.cu`. But, I don't see the macro defined in the master branch (`aten/src/ATen/Dispatch.h`). I'm not aware if there is a PR to add this macro.
* Updated tests. #37791 used `expectedFailureCUDA` in `torch/testing/_internal/common_methods_invocations.py` since there was no support for complex types (CUDA) for `tanh` by then. Removed the keyword from there.

cc: @anjali411

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu (hunk 4 above)
Comment (krshrimali) [reply aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:168]: As per offline (slack) discussion, this has been reverted back from using `c10::complex`. Eventually though, we'll use c10::complex once we have the support.

File: aten/src/ATen/native/cuda/UnaryGeometricKernels.cu (hunk 5 above)
Comment (krshrimali) [reply aten/src/ATen/native/cuda/UnaryGeometricKernels.cu:162]: Reverted back to `thrust::complex<T>` till the time full migration from `std::complex` to `c10::complex` is complete. :)

File: aten/src/ATen/native/cuda/BinaryMiscBackwardOpsKernels.cu [hunk 6]
```diff
@@ -20,7 +20,7 @@ void sigmoid_backward_kernel_cuda(TensorIterator& iter) {
 }
 
 void tanh_backward_kernel_cuda(TensorIterator& iter) {
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "tanh_backward_cuda", [&]() {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(iter.dtype(), "tanh_backward_cuda", [&]() {
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/BinaryMiscBackwardOpsKernels.cu:23]: `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.dtype() ...)`

**Comment by anjali411** [comment]
let's add the dispatch for tanh for Half back again and it should be good

File: aten/src/ATen/native/cuda/BinaryMiscBackwardOpsKernels.cu (hunk 6 above)
Comment (krshrimali) [reply aten/src/ATen/native/cuda/BinaryMiscBackwardOpsKernels.cu]: Should be resolved here: https://github.com/pytorch/pytorch/pull/38486/commits/a38b4b34a2a70d9498ae25cddb3199805072449a. Thanks!

File: aten/src/ATen/native/cuda/BinaryMiscBackwardOpsKernels.cu [hunk 7]
```diff
@@ -20,7 +20,7 @@ void sigmoid_backward_kernel_cuda(TensorIterator& iter) {
 }
 
 void tanh_backward_kernel_cuda(TensorIterator& iter) {
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "tanh_backward_cuda", [&]() {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.dtype(), "tanh_backward_cuda", [&]() {
     gpu_kernel(iter, []GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
```
Comment (anjali411) [inline aten/src/ATen/native/cuda/BinaryMiscBackwardOpsKernels.cu:24]: `using thrust_t = typename ztype_cuda<scalar_t>::thrust_t;`

Comment (krshrimali) [reply aten/src/ATen/native/cuda/BinaryMiscBackwardOpsKernels.cu]: Should be resolved here: https://github.com/pytorch/pytorch/pull/38486/commits/c36a4480497346ef3a831673a50be5d316645e33. Thanks!

**Comment by krshrimali** [comment]
Please take a look, @anjali411. There were a few merge conflicts, which I have resolved. The 2 failures reported by CI seem unrelated to the changes in this PR.

**Comment by krshrimali** [comment]
Closing this PR as another PR added complex support. Thanks!

**Pull request closed without merging**